From e2b66c52a6464cbf390acf2a1650f9b87bad2b0f Mon Sep 17 00:00:00 2001
From: Rong Bao <webmaster@csmantle.top>
Date: Thu, 30 Apr 2026 22:37:14 +0800
Subject: [PATCH 13/15] Bug 2023167 - Part 9: Move lul::StackImage out of stack
 on {loongarch64,riscv64}-linux. r=#profiler-reviewers!

The purpose of this patch is to work around huge frames on these
platforms. Clang has little support for `-fstack-clash-protection` on
LoongArch, while on RV64 this flag is not enabled in build config by
default.

For async sampling, we store the UniquePtr to a heap-allocated
lul::StackImage in SamplerThread, and pass the pointer around when doing
LUL backtraces.

For sync sampling, we allocate space for lul::StackImage in the thread
local storage, and pass the pointer into LUL backtracer.

This patch is organized so that only {loongarch64,riscv64}-linux are
affected, while other architectures should not be impacted.

Differential Revision: https://phabricator.services.mozilla.com/D297847
---
 .../profiler/core/platform-linux-android.cpp  |  4 +
 tools/profiler/core/platform.cpp              | 81 ++++++++++++++-----
 2 files changed, 63 insertions(+), 22 deletions(-)

diff --git a/tools/profiler/core/platform-linux-android.cpp b/tools/profiler/core/platform-linux-android.cpp
index 9029e0e4f94619cb8eff5b2fffc80a5498d7bc12..9463e196be8f1a69c158cdd12dc6efb8300bc7de 100644
--- a/tools/profiler/core/platform-linux-android.cpp
+++ b/tools/profiler/core/platform-linux-android.cpp
@@ -539,6 +539,10 @@ SamplerThread::SamplerThread(PSLockRef aLock, uint32_t aActivityGeneration,
   }
 #endif
 
+#if defined(USE_PREINIT_LUL_STACK_IMAGE)
+  mLulStackImage = MakeUnique<lul::StackImage>();
+#endif
+
   // Start the sampling thread. It repeatedly sends a SIGPROF signal. Sending
   // the signal ourselves instead of relying on itimer provides much better
   // accuracy.
diff --git a/tools/profiler/core/platform.cpp b/tools/profiler/core/platform.cpp
index b454b3fab0285a378ee4dc4605af770d7e399415..89244680dd469b1e70eaabe073af5500c137ff50 100644
--- a/tools/profiler/core/platform.cpp
+++ b/tools/profiler/core/platform.cpp
@@ -247,6 +247,13 @@ BOOL WINAPI GetThreadInformation(
 #  define HAVE_FASTINIT_NATIVE_UNWIND
 #endif
 
+// On these platforms, lul::StackImage is large, and the compiler support for
+// -fstack-clash-protection is poor. We need to alloc it beforehand.
+#if defined(USE_LUL_STACKWALK) && \
+    (defined(GP_PLAT_loongarch64_linux) || defined(GP_PLAT_riscv64_linux))
+#  define USE_PREINIT_LUL_STACK_IMAGE
+#endif
+
 #ifdef MOZ_VALGRIND
 #  include <valgrind/memcheck.h>
 #else
@@ -2758,7 +2765,8 @@ MOZ_ASAN_IGNORE static void ASAN_memcpy(void* aDst, const void* aSrc,
 static void DoLULBacktrace(
     const ThreadRegistration::UnlockedReaderAndAtomicRWOnThread& aThreadData,
     const Registers& aRegs, NativeStack& aNativeStack,
-    StackWalkControl* aStackWalkControlIfSupported) {
+    StackWalkControl* aStackWalkControlIfSupported,
+    lul::StackImage* aStackImg) {
   // WARNING: this function runs within the profiler's "critical section".
   // WARNING: this function might be called while the profiler is inactive, and
   //          cannot rely on ActivePS.
@@ -2837,7 +2845,7 @@ static void DoLULBacktrace(
   // function (DoNativeBacktrace)'s frame reasonable.  Most stacks observed in
   // practice are small, 4KB or less, and so the copy costs are insignificant
   // compared to other profiler overhead.
-  //
+#  if !defined(USE_PREINIT_LUL_STACK_IMAGE)
   // |stackImg| is allocated on this (the sampling thread's) stack.  That
   // implies that the frame for this function is at least N_STACK_BYTES large.
   // In general it would be considered unacceptable to have such a large frame
@@ -2848,8 +2856,16 @@ static void DoLULBacktrace(
   // thread safe, which would be a problem if we ever allow multiple sampler
   // threads.  Hence allocating it on the stack seems to be the least-worst
   // option.
-
-  lul::StackImage stackImg;
+  lul::StackImage stackImgOnStack;
+  lul::StackImage* const stackImg = &stackImgOnStack;
+#  else
+  // |stackImg| is allocated by the sampling thread, but on the heap.  This
+  // works around the potential guard page skip and the inability of Clang to
+  // to generate appropriate instructions for `-fstack-clash-protection` on
+  // some Tier-3 platforms.
+  MOZ_ASSERT(aStackImg != nullptr);
+  lul::StackImage* const stackImg = aStackImg;
+#  endif
 
   {
 #  if defined(GP_PLAT_amd64_linux) || defined(GP_PLAT_amd64_android) || \
@@ -2895,8 +2911,8 @@ static void DoLULBacktrace(
       }
     }
     MOZ_ASSERT(nToCopy <= lul::N_STACK_BYTES);
-    stackImg.mLen = nToCopy;
-    stackImg.mStartAvma = start;
+    stackImg->mLen = nToCopy;
+    stackImg->mStartAvma = start;
     if (nToCopy > 0) {
       // If this is a vanilla memcpy(), ASAN makes the following complaint:
       //
@@ -2908,11 +2924,11 @@ static void DoLULBacktrace(
       // This code is very much a custom stack unwind mechanism! So we use an
       // alternative memcpy() implementation that is ignored by ASAN.
 #  if defined(MOZ_HAVE_ASAN_IGNORE)
-      ASAN_memcpy(&stackImg.mContents[0], (void*)start, nToCopy);
+      ASAN_memcpy(&stackImg->mContents[0], (void*)start, nToCopy);
 #  else
-      memcpy(&stackImg.mContents[0], (void*)start, nToCopy);
+      memcpy(&stackImg->mContents[0], (void*)start, nToCopy);
 #  endif
-      (void)VALGRIND_MAKE_MEM_DEFINED(&stackImg.mContents[0], nToCopy);
+      (void)VALGRIND_MAKE_MEM_DEFINED(&stackImg->mContents[0], nToCopy);
     }
   }
 
@@ -2922,7 +2938,7 @@ static void DoLULBacktrace(
   lul->Unwind(reinterpret_cast<uintptr_t*>(aNativeStack.mPCs),
               reinterpret_cast<uintptr_t*>(aNativeStack.mSPs),
               &aNativeStack.mCount, &framePointerFramesAcquired,
-              MAX_NATIVE_FRAMES, &startRegs, &stackImg);
+              MAX_NATIVE_FRAMES, &startRegs, stackImg);
 
   // Update stats in the LUL stats object.  Unfortunately this requires
   // three global memory operations.
@@ -2937,15 +2953,15 @@ static void DoLULBacktrace(
 static void DoNativeBacktrace(
     const ThreadRegistration::UnlockedReaderAndAtomicRWOnThread& aThreadData,
     const Registers& aRegs, NativeStack& aNativeStack,
-    StackWalkControl* aStackWalkControlIfSupported) {
+    StackWalkControl* aStackWalkControlIfSupported, void* aStackImg = nullptr) {
   // This method determines which stackwalker is used for periodic and
   // synchronous samples. (Backtrace samples are treated differently, see
   // profiler_suspend_and_sample_thread() for details). The only part of the
   // ordering that matters is that LUL must precede FRAME_POINTER, because on
   // Linux they can both be present.
 #  if defined(USE_LUL_STACKWALK)
-  DoLULBacktrace(aThreadData, aRegs, aNativeStack,
-                 aStackWalkControlIfSupported);
+  DoLULBacktrace(aThreadData, aRegs, aNativeStack, aStackWalkControlIfSupported,
+                 reinterpret_cast<lul::StackImage*>(aStackImg));
 #  elif defined(USE_EHABI_STACKWALK)
   DoEHABIBacktrace(aThreadData, aRegs, aNativeStack,
                    aStackWalkControlIfSupported);
@@ -3037,7 +3053,8 @@ static inline void DoSharedSample(
     const ThreadRegistration::UnlockedReaderAndAtomicRWOnThread& aThreadData,
     JsFrame* aJsFrames, const Registers& aRegs, uint64_t aSamplePos,
     uint64_t aBufferRangeStart, ProfileBuffer& aBuffer,
-    StackCaptureOptions aCaptureOptions = StackCaptureOptions::Full) {
+    StackCaptureOptions aCaptureOptions = StackCaptureOptions::Full,
+    void* aStackImg = nullptr) {
   // WARNING: this function runs within the profiler's "critical section".
 
   MOZ_ASSERT(!aBuffer.IsThreadSafe(),
@@ -3063,7 +3080,7 @@ static inline void DoSharedSample(
 #if defined(HAVE_NATIVE_UNWIND)
   if (captureNative) {
     DoNativeBacktrace(aThreadData, aRegs, nativeStack,
-                      stackWalkControlIfSupported);
+                      stackWalkControlIfSupported, aStackImg);
 
     MergeStacks(aIsSynchronous, aThreadData, nativeStack, collector, aJsFrames,
                 jsFramesCount);
@@ -3085,7 +3102,7 @@ static void DoSyncSample(
     uint32_t aFeatures,
     const ThreadRegistration::UnlockedReaderAndAtomicRWOnThread& aThreadData,
     const TimeStamp& aNow, const Registers& aRegs, ProfileBuffer& aBuffer,
-    StackCaptureOptions aCaptureOptions) {
+    StackCaptureOptions aCaptureOptions, void* aStackImg = nullptr) {
   // WARNING: this function runs within the profiler's "critical section".
 
   MOZ_ASSERT(aCaptureOptions != StackCaptureOptions::NoStack,
@@ -3103,7 +3120,7 @@ static void DoSyncSample(
     // No JSContext, there is no JS frame buffer (and no need for it).
     DoSharedSample(/* aIsSynchronous = */ true, aFeatures, aThreadData,
                    /* aJsFrames = */ nullptr, aRegs, samplePos,
-                   bufferRangeStart, aBuffer, aCaptureOptions);
+                   bufferRangeStart, aBuffer, aCaptureOptions, aStackImg);
   } else {
     // JSContext is present, we need to lock the thread data to access the JS
     // frame buffer.
@@ -3114,7 +3131,7 @@ static void DoSyncSample(
             DoSharedSample(/* aIsSynchronous = */ true, aFeatures, aThreadData,
                            aLockedThreadData.GetJsFrameBuffer(), aRegs,
                            samplePos, bufferRangeStart, aBuffer,
-                           aCaptureOptions);
+                           aCaptureOptions, aStackImg);
           });
     });
   }
@@ -3127,7 +3144,7 @@ static inline void DoPeriodicSample(
     PSLockRef aLock,
     const ThreadRegistration::UnlockedReaderAndAtomicRWOnThread& aThreadData,
     const Registers& aRegs, uint64_t aSamplePos, uint64_t aBufferRangeStart,
-    ProfileBuffer& aBuffer) {
+    ProfileBuffer& aBuffer, void* aStackImg = nullptr) {
   // WARNING: this function runs within the profiler's "critical section".
 
   MOZ_RELEASE_ASSERT(ActivePS::Exists(aLock));
@@ -3135,7 +3152,7 @@ static inline void DoPeriodicSample(
   JsFrameBuffer& jsFrames = CorePS::JsFrames(aLock);
   DoSharedSample(/* aIsSynchronous = */ false, ActivePS::Features(aLock),
                  aThreadData, jsFrames, aRegs, aSamplePos, aBufferRangeStart,
-                 aBuffer);
+                 aBuffer, StackCaptureOptions::Full, aStackImg);
 }
 
 #undef UNWINDING_REGS_HAVE_ECX_EDX
@@ -4536,6 +4553,10 @@ class SamplerThread {
   // The interval between samples, measured in microseconds.
   const int mIntervalMicroseconds;
 
+#if defined(USE_PREINIT_LUL_STACK_IMAGE)
+  UniquePtr<lul::StackImage> mLulStackImage;
+#endif
+
   // The OS-specific handle for the sampler thread.
 #if defined(GP_OS_windows)
   HANDLE mThread;
@@ -4955,7 +4976,13 @@ void SamplerThread::Run() {
                   [&](const Registers& aRegs, const TimeStamp& aNow) {
                     DoPeriodicSample(lock, lockedThreadData.DataCRef(), aRegs,
                                      samplePos, bufferRangeStart,
-                                     localProfileBuffer);
+                                     localProfileBuffer,
+#if defined(USE_PREINIT_LUL_STACK_IMAGE)
+                                     mLulStackImage.get()
+#else
+                                     nullptr
+#endif
+                    );
 
                     // For "eventDelay", we want the input delay - but if
                     // there are no events in the input queue (or even if there
@@ -7884,6 +7911,10 @@ bool profiler_capture_backtrace_into(ProfileChunkedBuffer& aChunkedBuffer,
 
         ProfileBuffer profileBuffer(aChunkedBuffer);
 
+#if defined(USE_PREINIT_LUL_STACK_IMAGE)
+        static thread_local lul::StackImage sLulStackImage;
+#endif
+
         Registers regs;
 #if defined(HAVE_NATIVE_UNWIND)
         REGISTERS_SYNC_POPULATE(regs);
@@ -7893,7 +7924,13 @@ bool profiler_capture_backtrace_into(ProfileChunkedBuffer& aChunkedBuffer,
 
         DoSyncSample(*maybeFeatures,
                      aOnThreadRef.UnlockedReaderAndAtomicRWOnThreadCRef(),
-                     TimeStamp::Now(), regs, profileBuffer, aCaptureOptions);
+                     TimeStamp::Now(), regs, profileBuffer, aCaptureOptions,
+#if defined(USE_PREINIT_LUL_STACK_IMAGE)
+                     &sLulStackImage
+#else
+                     nullptr
+#endif
+        );
 
         return true;
       },
-- 
2.55.0

